mypy/solve.py

```diff
@@ -33,62 +41,332 @@ def solve_constraints(
 
     res: list[Type | None] = []
+    for v in vars:
+        if v in solutions:
+            res.append(solutions[v])
+        else:
+            # No constraints for type variable -- 'UninhabitedType' is the most specific type.
+            candidate: Type
+            if strict:
+                candidate = UninhabitedType()
+                candidate.ambiguous = True
+            else:
+                candidate = AnyType(TypeOfAny.special_form)
+            res.append(candidate)
+    return res
+
+
+def solve_non_linear(
+    vars: list[TypeVarId], constraints: list[Constraint], cmap: dict[TypeVarId, list[Constraint]]
+) -> dict[TypeVarId, Type | None]:
+    """Solve set of constraints that may include non-linear ones, like T <: List[S].
 
-    # Solve each type variable separately.
+    The whole algorithm consists of five steps:
+      * Propagate via linear constraints to get all possible constraints for each variable
+      * Find dependencies between type variables, group them in SCCs, and sor topologically
+      * Check all SCC are intrinsically linear, we can't solve (express) T <: List[T]
+      * Variables in leaf SCCs that don't have constant bounds are free (choose one per SCC)
+      * Solve constraints iteratively starting from leafs, updating targets after each step.
+    """
+    extra_constraints = []
     for tvar in vars:
-        bottom: Type | None = None
-        top: Type | None = None
-        candidate: Type | None = None
-
-        # Process each constraint separately, and calculate the lower and upper
-        # bounds based on constraints. Note that we assume that the constraint
-        # targets do not have constraint references.
-        for c in cmap.get(tvar, []):
-            if c.op == SUPERTYPE_OF:
-                if bottom is None:
-                    bottom = c.target
-                else:
-                    if type_state.infer_unions:
-                        # This deviates from the general mypy semantics because
-                        # recursive types are union-heavy in 95% of cases.
-                        bottom = UnionType.make_union([bottom, c.target])
-                    else:
-                        bottom = join_types(bottom, c.target)
+        extra_constraints.extend(propagate_constraints_for(tvar, SUBTYPE_OF, cmap))
+        extra_constraints.extend(propagate_constraints_for(tvar, SUPERTYPE_OF, cmap))
+    constraints += remove_dups(extra_constraints)
+
+    # Recompute constraint map after propagating.
+    cmap = {tv: [] for tv in vars}
+    for con in constraints:
+        if con.type_var in vars:
+            cmap[con.type_var].append(con)
+
+    dmap = compute_dependencies(cmap)
+    sccs = list(strongly_connected_components(set(vars), dmap))
+    if all(check_linear(scc, cmap) for scc in sccs):
+        raw_batches = list(topsort(prepare_sccs(sccs, dmap)))
+        leafs = raw_batches[0]
+        free_vars = []
+        for scc in leafs:
+            if all(
```

Comment:
Add comment here about the purpose of this if statement (I think I figured it out but it wasn't obvious).